nipype/pipeline/engine/nodes.py

```diff
@@ -293,27 +293,29 @@ def is_cached(self, rm_outdated=False):
         """
         outdir = self.output_dir()
 
-        # Update hash
-        hashed_inputs, hashvalue = self._get_hashval()
-
         # The output folder does not exist: not cached
-        if not op.exists(outdir):
-            logger.debug('[Node] Directory not found "%s".', outdir)
+        if not op.exists(outdir) or \
+                not op.exists(op.join(outdir, 'result_%s.pklz' % self.name)):
+            logger.debug('[Node] Not cached "%s".', outdir)
             return False, False
 
-        hashfile = op.join(outdir, '_0x%s.json' % hashvalue)
-        cached = op.exists(hashfile)
-
-        # Check if updated
+        # Check if there are hashfiles
         globhashes = glob(op.join(outdir, '_0x*.json'))
         unfinished = [
             path for path in globhashes
             if path.endswith('_unfinished.json')
         ]
         hashfiles = list(set(globhashes) - set(unfinished))
+
+        # Update hash
+        hashed_inputs, hashvalue = self._get_hashval()
+
+        hashfile = op.join(outdir, '_0x%s.json' % hashvalue)
         logger.debug('[Node] Hashes: %s, %s, %s, %s',
                      hashed_inputs, hashvalue, hashfile, hashfiles)
 
+        cached = op.exists(hashfile)
```

Comment:
Can `hashfile` be created by `self._get_hashval()`? If not, we can avoid hitting the filesystem again:

```suggestion
        cached = hashfile in hashfiles
```

Reply:
Creating the `hashfile` within `self._get_hashval` seems a good option. I'll send a different PR, so that this one involves the minimal set of changes.

Reply:
Is there no case in which we might want to get the hash without generating a file?

Reply:
yes, often we may want to do this. so i wouldn't create the file within that call.

Reply:
Looks like this got reverted in the force push.